Refactor to decouple request/response from the rest of identity

GarethCOliver opened the pull request:
This is a pure refactor that moves request/response out of identity into their own package. It also breaks references from identity. To do that a 'shared' package is created for common utilities that are independent of identity.

The goal here is to make it easier for a server side implementations to dependent on request generation and response parsing without having to pull in the rest of identity which is still a bit android specific.

davidz25 commented:
Timing-wise, @suzannajiwani is working on publishing Javadoc (see PR #267 and PR #268) and pushing releases out to GMaven. I think we should probably land that first so you might need to do an annoying rebase especially with some of the Javadoc cleanup. We could probably have timed this better.

Back to the PR, I think your change makes sense but I was also expecting to see changes to the build files to generate two libraries, one only for Android apps and another one that can be used for any JVM environment. We'll also want to push both artifacts to GMaven but we can revisit that when we've done the first release.

Thoughts?

GarethCOliver commented:
That SGTM, happy to wait till after. 

Yeah, that's the end goal, this one here just provides encapsulation between this and identity. There is still a bit of work needed to break all dependencies on android entirely (some androidx annotations currently, and also the Logger needs to be injectable as it has a dependency on logcat). At that point it should probably be pulled out into its own java library (don't think it should need to be both android and java, right?). Was intending to do it as a follow-ups but if you'd prefer I can do it as a single large one.

GarethCOliver commented:
Rebased on top of the changes, and added an internal.common package for the utilities that have broadened visibility. We might need to re-think a bit how to do dependencies if request/response gets more pulled out to make them java-only, but this at least preserves the visibility limits for now (and, probably, useful to have more things in internal in general to make the public surface more clear).

davidz25 commented:
I like this approach and I've been wanting to reorganize things in packages especially since we're going to be landing a new credential storage framework which will work both on generic JVM and on Android (see https://github.com/google/identity-credential/tree/keystore-abstraction branch).

I'm thinking the packages we want are the following

- com.android.identity
  - util
    - Would have Logger, Constants, Utility, etc.
      - (the Constants and Utility classes are ugly and anti-patterns will likely be removed in the future once we find good homes for what they currently do.)
  - internal
    - would have Util.java and CborUtil.java etc.
      - Again, those are pretty ugly and will likely be replaced with better designed code but this is less urgent as it's marked as internal via the package name
        - for example we're looking at depending on https://github.com/android/cose-lib in the near future
  - mdoc
    - Would have Device{Request, Respect}{Parser, Generator}, Engagement{Parser, Generator}, MobileSecurityObject{Parser, Generator}, OriginInfo*
  - storage
    - Would have all the generic stuff from keystore-abstraction branch
  - android
    - storage
      - Would have all the Android-specific bits from keystore-abstraction branch
    - mdoc
      - Would have DataTransport*, DeviceRetrievalHelper, NfcEngagementHelper, QrEngagementHelper etc

Thoughts? Concretely I think this PR could move just the classes indicated above in the `mdoc`, `util`, and `internal` packages and we can move the rest in later PRs. It's a little bit more work than you done here but I hope it's not too much!

Cc @suzannajiwani for visibility.

GarethCOliver commented:
Updated. Haven't moved Engagement{Parser, Generator} due to it having hard dependencies on ConnectionMethod + DataTransport which will require a bit more thought for abstraction.

I did also add a LogPrinter to Logger to extract out the logcat logging, as well as allow other sources to implement their own log printing. Using ApplicationContext as the place to set it up for the two apps. This lets the Logger class also remove its android dependencies.

davidz25 commented on identity/src/main/java/com/android/identity/mdoc/response/MobileSecurityObjectGenerator.java:
```diff
@@ -14,11 +14,14 @@
  */
 
-package com.android.identity;
+package com.android.identity.mdoc.response;
```
I don't think `mdoc.response` is the right place because it's also something you need on the issuer side... `mdoc.mso` maybe?

davidz25 replied:
Also, the tests should be moved as well

GarethCOliver replied:
done

davidz25 commented on identity/src/main/java/com/android/identity/android/util/AndroidLogPrinter.kt:
```diff
@@ -0,0 +1,20 @@
+import com.android.identity.util.Logger.LEVEL_W
+import com.android.identity.util.Logger.LogPrinter
+
+/** An implementation of [LogPrinter] that logs to logcat. */
```
is `[LogPrinter]` something Kotlin specific? The Java way to do this is `{@link LogPrinter}`

GarethCOliver replied:
Yeah, kotlin uses '[...]' rather than {@link ...} Now that we are using dokka this also works, yay! https://kotlinlang.org/docs/kotlin-doc.html#links-to-elements

davidz25 commented on appholder/src/main/java/com/android/mdl/app/MainActivity.kt:
```diff
@@ -107,7 +105,12 @@ class MainActivity : AppCompatActivity() {
+                mdocReferrerUri
+              )
+            )
         }
```
Seems like unintentional noise being introduced? (Couple of other places too)

davidz25 replied:
Would be nice to fix this...

GarethCOliver replied:
Fixed, unintentional noise from the refactor. Ran fixes over the code base which did a few other things in addition to cleaning up some of the cruft (added finals, fixed asserts etc).

davidz25 commented on identity/src/main/java/com/android/identity/Utility.java:
```diff
@@ -18,19 +18,32 @@
 
```
I was going to ask if we move this entire file to `android/util/AndroidUtil.java` because I think only Android stuff currently depends on it... but I think we can also just do that in a later PR.

GarethCOliver replied:
ack

davidz25 commented on identity/src/main/java/com/android/identity/internal/CborUtil.java:
```diff
@@ -1,10 +1,11 @@
-package com.android.identity;
+package com.android.identity.internal;
```
So `Util.java` is a hot mess but I think `CborUtil` is actually fine and could be moved to the `util` package so apps can use it too.

GarethCOliver replied:
done

davidz25 reviewed (commented):
A couple of inline comments.

Also, I think we also want this split into two different libraries, yes? I fear if we don't do that initially we'll just end up putting Android-specific code into where it shouldn't go which will break other users. Thoughts?

davidz25 commented on appholder/src/main/java/com/android/mdl/app/MainActivity.kt:
```diff
@@ -101,13 +99,18 @@ class MainActivity : AppCompatActivity() {
         logInfo("uri: $mdocUri")
 
-        val originInfos = ArrayList<OriginInfo>()
+        val originInfos = ArrayList<com.android.identity.mdoc.origininfo.OriginInfo>()
```
The `com.android.identity.mdoc.origininfo.` part seems unnecessary since you import this above? There's a bunch of places where this happen, I guess a side-effect of refactoring? Either way, would be nice to clean it up before landing.

davidz25 reviewed (changes requested):
This looks great! Would appreciate if we can clean up the noise in the callsites (appholder and appverifier) before landing - it shouldn't take more than a couple of minutes to fix that up! Let's see if we can land this today!

davidz25 reviewed (changes requested):
This looks great! Would appreciate if we can clean up the noise in the callsites (appholder and appverifier) before landing - it shouldn't take more than a couple of minutes to fix that up! Let's see if we can land this today!

GarethCOliver commented:
Should be all cleaned up and ready for review. There might need to be some tweaking to make the publish work, but probably better to do in a follow-up as it'll be small build tweaks.

davidz25 reviewed (approved):
LGTM, thanks for this huge change!
